<a href="https://colab.research.google.com/github/luisdh8/2049-FokusCurso/blob/main/Utilizando_chatgpt_como_asistente.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <strong><font color="7b71ff">Análisis de datos con python: Utilizando ChatGPT como asistente</font></strong>


---

<img src="https://raw.githubusercontent.com/alura-cursos/ia-datascience-dataviz/main/imagens/logos/logo_branca_fundo_azul.png" alt="inserir alt">

# **Aula 1 - Conociendo los datos**

 Zoop es una gran minorista que atiende a todas las regiones de México a través de su e-commerce. Es conocida por la amplia variedad de productos que ofrece, buscando satisfacer a todo tipo de público. Posee una serie de centros de distribución (CDs) para la entrega de sus productos, la mayoría de ellos situados en el estado de Mexico. También posee una cadena de supermercados, y vamos a trabajar como científicos de datos para esta prestigiosa empresa.

### **Accesando a los datos**

In [62]:
url = 'https://gist.githubusercontent.com/ahcamachod/4b964953032a2078d67052ec8fb92e72/raw/8106d7c3f001390744ec3aa657b9847f306c9df8/zoop_supermercado.json'

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

Vamos a actuar como científicos de datos en Zoop.

Recibimos datos en formato de archivo JSON, estructurados en listas y diccionarios. El archivo está disponible en una URL de Github.

Necesitamos analizar esos datos para ayudar en las decisiones de la empresa. Programamos en Python y conocemos las bibliotecas de análisis de datos Numpy y Pandas. Por lo tanto, prioriza el uso de estas bibliotecas.

Igualmente, infórmanos cómo podemos leer estos datos en formato JSON y para poder visualizarlos en formato de tabla en Python.

In [63]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)

# Leer directamente el JSON desde la URL
df = pd.read_json(url)

In [64]:
df.head()

item_identificador tienda_identificador  ventas_totales  \
0              FDB08               OUT018       176503.58   
1              DRQ35               OUT049       185758.20   
2              FDD14               OUT018       165983.94   
3              FDY37               OUT045       314923.40   
4              FDY59               OUT018        64782.34   

                                                item  \
0  {'item_peso': 6.055, 'item_contenido_grasa': '...   
1  {'item_peso': 9.3, 'item_contenido_grasa': 'Ba...   
2  {'item_peso': 20.7, 'item_contenido_grasa': 'B...   
3  {'item_peso': 17.0, 'item_contenido_grasa': 'R...   
4  {'item_peso': 8.195, 'item_contenido_grasa': '...   

                                              tienda  
0  {'tienda_ano_establecimiento': 2019, 'tienda_t...  
1  {'tienda_ano_establecimiento': 2009, 'tienda_t...  
2  {'tienda_ano_establecimiento': 2019, 'tienda_t...  
3  {'tienda_ano_establecimiento': 2012, 'tienda_t...  
4  {'tienda_ano_establecimiento': 2019, 'tienda_t...

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>


El DataFrame "df" posee las columnas "item_identificador", "tienda_identificador", "ventas_totales", "item" y "tienda".

Las columnas "item" y "tienda" contienen diccionarios anidados dentro de cada una de las filas. ¿Cómo puedo transformar los datos de los diccionarios y columnas para hacerlos más accesibles en mi DataFrame?


In [65]:
# Expande la columna 'item' y convierte los datos JSON en columnas
item_df = pd.json_normalize(df['item'])
# Expande la columna 'tienda' y convierte los datos JSON en columnas
tienda_df = pd.json_normalize(df['tienda'])

In [66]:
# Concatenar las columnas expandidas al DataFrame original
df = pd.concat([df.drop(['item', 'tienda'], axis=1), item_df, tienda_df], axis=1)

In [67]:
# Mostrar las primeras filas del DataFrame resultante
df.head()

item_identificador tienda_identificador  ventas_totales  item_peso  \
0              FDB08               OUT018       176503.58      6.055   
1              DRQ35               OUT049       185758.20      9.300   
2              FDD14               OUT018       165983.94     20.700   
3              FDY37               OUT045       314923.40     17.000   
4              FDY59               OUT018        64782.34      8.195   

      item_contenido_grasa  item_visibilidad            item_tipo  \
0  Bajo Contenido de Grasa          0.031230   Frutas y Vegetales   
1  Bajo Contenido de Grasa          0.042357  Bebidas Alcohólicas   
2  Bajo Contenido de Grasa          0.170500            Enlatados   
3                  Regular          0.026623            Enlatados   
4  Bajo Contenido de Grasa          0.000000             Dulcería   

   item_precio item_cantidad_venta  tienda_ano_establecimiento tienda_tamano  \
0       160.36                None                        2019         Medio   
1       123.24                None                        2009         Medio   
2       184.13                None                        2019         Medio   
3       144.25                None                        2012          None   
4        93.15                None                        2019         Medio   

  tienda_tipo_ubicacion          tienda_tipo  
0               Nivel 3  Supermercado Tipo 2  
1               Nivel 1  Supermercado Tipo 1  
2               Nivel 3  Supermercado Tipo 2  
3               Nivel 2  Supermercado Tipo 1  
4               Nivel 3  Supermercado Tipo 2

# **Clase 2 - Exploración de los dados**

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

El DataFrame "df" contiene datos referentes a las ventas de diversos productos en varias tiendas del supermercado. Se proporcionan los IDs de productos y los IDs de tiendas que comercializaron esos productos, el total recaudado en ventas, así como descripciones sobre el producto vendido y la tienda que lo vendió.

Estamos iniciando la etapa de exploración de datos, en la cual deseamos conocer los valores informados e identificar si existen valores ausentes, duplicados y otras inconsistencias.

Construye un código para ejecutar la etapa de exploración de datos usando Python, comentando el paso a paso de manera clara y objetiva.

No es necesario utilizar "print()" para mostrar el resultado de las consultas.

In [68]:
# 1. Ver primeras filas del DataFrame para tener una idea general del contenido
df.head()

item_identificador tienda_identificador  ventas_totales  item_peso  \
0              FDB08               OUT018       176503.58      6.055   
1              DRQ35               OUT049       185758.20      9.300   
2              FDD14               OUT018       165983.94     20.700   
3              FDY37               OUT045       314923.40     17.000   
4              FDY59               OUT018        64782.34      8.195   

      item_contenido_grasa  item_visibilidad            item_tipo  \
0  Bajo Contenido de Grasa          0.031230   Frutas y Vegetales   
1  Bajo Contenido de Grasa          0.042357  Bebidas Alcohólicas   
2  Bajo Contenido de Grasa          0.170500            Enlatados   
3                  Regular          0.026623            Enlatados   
4  Bajo Contenido de Grasa          0.000000             Dulcería   

   item_precio item_cantidad_venta  tienda_ano_establecimiento tienda_tamano  \
0       160.36                None                        2019         Medio   
1       123.24                None                        2009         Medio   
2       184.13                None                        2019         Medio   
3       144.25                None                        2012          None   
4        93.15                None                        2019         Medio   

  tienda_tipo_ubicacion          tienda_tipo  
0               Nivel 3  Supermercado Tipo 2  
1               Nivel 1  Supermercado Tipo 1  
2               Nivel 3  Supermercado Tipo 2  
3               Nivel 2  Supermercado Tipo 1  
4               Nivel 3  Supermercado Tipo 2

In [71]:
# 2. Ver estructura general del DataFrame: columnas, tipos de datos, valores no nulos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8550 entries, 0 to 8549
Data columns (total 13 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   item_identificador          8550 non-null   object 
 1   tienda_identificador        8550 non-null   object 
 2   ventas_totales              8550 non-null   float64
 3   item_peso                   7081 non-null   float64
 4   item_contenido_grasa        8550 non-null   object 
 5   item_visibilidad            8550 non-null   float64
 6   item_tipo                   8550 non-null   object 
 7   item_precio                 8550 non-null   float64
 8   item_cantidad_venta         0 non-null      object 
 9   tienda_ano_establecimiento  8550 non-null   int64  
 10  tienda_tamano               6133 non-null   object 
 11  tienda_tipo_ubicacion       8550 non-null   object 
 12  tienda_tipo                 8550 non-null   object 
dtypes: float64(4), int64(1), object(8

In [70]:
# 3. Estadísticas descriptivas para columnas numéricas
df.describe()

ventas_totales    item_peso  item_visibilidad  item_precio  \
count    8.550000e+03  7081.000000       8550.000000  8550.000000   
mean     2.181949e+05    12.855023          0.066150   141.007453   
std      1.708098e+05     4.643508          0.051578    62.333062   
min      3.329000e+03     4.555000          0.000000    31.290000   
25%      8.349132e+04     8.770000          0.027024    93.787500   
50%      1.794331e+05    12.600000          0.053978   142.935000   
75%      3.100963e+05    16.850000          0.094646   185.760000   
max      1.308696e+06    21.350000          0.328391   266.890000   

       tienda_ano_establecimiento  
count                 8550.000000  
mean                  2007.830409  
std                      8.372541  
min                   1995.000000  
25%                   1997.000000  
50%                   2009.000000  
75%                   2014.000000  
max                   2019.000000

In [72]:
# 5. Verificar si existen valores ausentes (NaN) en alguna columna
df.isnull().sum()

item_identificador               0
tienda_identificador             0
ventas_totales                   0
item_peso                     1469
item_contenido_grasa             0
item_visibilidad                 0
item_tipo                        0
item_precio                      0
item_cantidad_venta           8550
tienda_ano_establecimiento       0
tienda_tamano                 2417
tienda_tipo_ubicacion            0
tienda_tipo                      0
dtype: int64

In [73]:
# 7. Verificar si hay filas duplicadas
df.duplicated().sum()

np.int64(27)

In [74]:
df.nunique()

item_identificador            1559
tienda_identificador            10
ventas_totales                3493
item_peso                      415
item_contenido_grasa             5
item_visibilidad              7880
item_tipo                       16
item_precio                   4839
item_cantidad_venta              0
tienda_ano_establecimiento       9
tienda_tamano                    3
tienda_tipo_ubicacion            3
tienda_tipo                      4
dtype: int64

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

El DataFrame "df" contiene datos relacionados con las ventas durante un trimestre de un supermercado. Deseo utilizarlo para verificar el desempeño de esas ventas.

Durante la exploración del "df", identificamos algunas inconsistencias, como:

* Filas duplicadas,
* Valores nulos y
* Datos categóricos que fueron escritos de manera no estandarizada, dando la idea de que existen más categorías de las que realmente existen.

¿Qué debo hacer para garantizar que esas inconsistencias que encontré no afecten mi análisis?

# **Clase 3 - Limpieza y tratamiento de los datos**

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>


Durante la exploración del DataFrame "df", identificamos 27 filas duplicadas usando "df.duplicated().sum()".

El objetivo es analizar los datos en "df" para medir el desempeño de una empresa durante un trimestre de ventas.

¿Qué debemos hacer con los datos duplicados para garantizar que no haya impacto negativo en el análisis?

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>


Identificamos tres columnas en "df" que presentan valores nulos:

* La columna "item_cantidad_venta" tiene 8550 filas con valores nulos.
* La columna "item_peso" tiene 1469 filas con valores nulos.
* La columna "tienda_tamano" tiene 2417 filas con valores nulos.

Hubo un gran número de ventas de artículos durante el trimestre, y no hay justificación para la ausencia de datos en la columna "item_cantidad_venta", lo que la convierte en una columna prescindible.

La columna "item_peso" contiene valores del tipo float, representando el peso en gramos de los productos.

La columna "tienda_tamano" tiene tres categorías de tamaño para supermercados.

¿Cómo debo manejar estos datos nulos para garantizar que mi análisis no se vea afectado?

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

La columna "item_conteudo_grasa" del DataFrame "df" presenta 5 valores únicos dados por: 'Bajo Contenido de Grasa', 'Regular', 'BTG', 'reg' y 'bajo contenido de grasa'.

Los valores únicos de la columna "item_contenido_grasa" deberían ser solo 'Bajo Contenido de Grasa' y 'Regular', pero fueron insertados sin estandarización, lo que resultó en más categorías de las deseadas.

¿Cómo puedo ajustar los datos de la columna "item_contenido_grasa" para mantener solo las dos categorías estándar?

# **Clase 4 - Visualización de datos**

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

Los datos del DataFrame "df" están limpios y tratados. Ahora, deseo iniciar la etapa de construcción de gráficos para crear visualizaciones que ofrezcan un buen soporte al análisis del desempeño de ventas de la empresa.

Las columnas disponibles en "df" son las siguientes:
- "item_identificador": Número único asignado a cada ítem
- "tienda_identificador": Nombre único de la tienda
- "ventas_totales": Valor obtenido en ventas totales
- "item_peso": Peso del ítem en gramos
- "item_contenido_grasa": Contenido de grasa del ítem
- "item_visibilidad": Valor de posicionamiento de cada ítem
- "item_tipo": Tipo de utilidad del ítem
- "item_precio": Precio unitario del ítem
- "tienda_ano_establecimiento": Año de Establecimiento de la tienda
- "tienda_tamano": Tamaño de la tienda
- "tienda_tipo_ubicacion": Tipo de ubicación de la tienda
- "tienda_tipo": Tipo de tienda

El objetivo es encontrar patrones o información que se destaquen cuando nos referimos al total obtenido en ventas especificado en la columna "ventas_totales".

Con base en la información proporcionada, ¿qué tipos de visualizaciones serían útiles para alcanzar el objetivo de analizar el desempeño de ventas?

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

El DataFrame "df" contiene la columna "ventas_totales", que informa el valor en reales obtenido en ventas de productos por varias tiendas.

Además, en "df", tenemos la columna "tienda_tamano" con 4 categorías de tamaño para tiendas: 'Grande', 'Medio', 'Pequeño' y 'No registrado'.

Deseo crear una visualización que permita identificar el total obtenido en ventas para las cuatro categorías de tamaño de tienda y representarlas en un gráfico de barras.

Los datos en la columna "ventas_totales" necesitan ser agrupados por las categorías de "tienda_tamano".

La figura del gráfico debe seguir la identidad visual de la tienda, con el fondo de la figura y del gráfico en color "#ebebeb" y las barras en color "#e23155".

Además, es necesario mostrar el total de las ventas en millones encima de cada columna, con el texto de esta información en color "#19325c" y la estructura "MXN $ valor M".

Utiliza las bibliotecas de visualización de datos de Python para crear esta visualización.

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>


El DataFrame "df" contiene la columna "ventas_totales", que informa el valor en reales obtenido en ventas de productos por varias tiendas.

Además, en "df", tenemos la columna "tienda_tipo" con 4 categorías de tipo para tiendas: 'Supermercado Tipo 1', 'Supermercado Tipo 2', 'Mercado', 'Supermercado Tipo 3'.

Deseo crear una visualización que permita identificar el total obtenido en ventas para las cuatro categorías de tipo de tienda y representarlas en un gráfico de barras.

Los datos en la columna "ventas_totales" deben ser agrupados por las categorías de "tienda_tipo".

La figura de la gráfica debe seguir la identidad visual de la tienda, con el fondo de la figura y de la gráfica en color "#ebebeb" y las barras en color "#e23155".

Además, es necesario mostrar el total de las ventas en millones sobre cada columna, con el texto de esa información en color "#19325c" y la estructura "MXN$ valor M".

Utiliza las bibliotecas de visualización de datos de Python para crear esta visualización.

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

El DataFrame "df" contiene la columna "ventas_totales", que informa el valor en reales obtenido en ventas de productos por varias tiendas.

Además, en "df", tenemos la columna "item_tipo" con 16 categorías de productos.

Deseo crear una visualización que permita identificar el total obtenido en ventas por categorías de productos en un gráfico de barras horizontales.

Los datos en la columna "ventas_totales" deben agruparse por las categorías de "item_tipo".

La figura de la representación debe seguir la identidad visual de la tienda, con el fondo de la figura y del gráfico en el color "#ebebeb".

Las barras deben estar en el color "#e23155".

Los valores de las barras deben estar ordenados para mantener un aspecto de embudo.

Además, es necesario mostrar el total de las ventas en millones dentro o al lado de cada barra, con el texto de esa información en el color "#19325c" y en la estructura "MXN$ valor M".

No muestres la cuadrícula en el gráfico.

Utiliza las bibliotecas de visualización de datos de Python para crear esta visualización.